pennylane/templates/layer.py

```diff
@@ -29,10 +29,13 @@ def _preprocess(args, depth):
     """
 
     for arg in args:
-        # TODO: handle ragged arrays without warnings
-        if shape(arg)[0] != depth:
+        try:
+            arg_depth = len(arg)
+        except TypeError:
+            arg_depth = shape(arg)[0]
```

Comment:
It is not inconceivable that there is some configuration of `arg` that is ragged, fails to return `len` and also is cast to a `numpy` array when calling `shape`, and that will still fail here. 

But I don't know what that configuration of `arg` is, and all our tests pass with this.

Reply:
what kinds of things fail to return from `len()` but can be cast to numpy arrays and return shapes? Curious why this `except` section is needed

Reply:
https://github.com/PennyLaneAI/pennylane/actions/runs/6125094486/job/16626384949

Reply:
if we only think TF will land us in this except block, then I'm happy with this, especially because TF won't allow ragged arrays either. maybe we can use `hasattr` to avoid `try-except`? eg.
```python
arg_depth = len(arg) if hasattr(arg, "__len__") else shape(arg)[0]
```
and if `shape(arg)` fails, then it's probably the calling object's fault : )